File: app/src/androidTest/java/com/nearsoft/nearbooks/ApplicationTest.java
package com.nearsoft.nearbooks;

import android.os.SystemClock;
import android.test.ApplicationTestCase;

import com.nearsoft.nearbooks.di.components.NearbooksApplicationComponent;
import com.nearsoft.nearbooks.models.BookModel;
import com.nearsoft.nearbooks.models.sqlite.Book;
import com.nearsoft.nearbooks.models.sqlite.Borrow;
import com.nearsoft.nearbooks.util.ErrorUtil;
import com.nearsoft.nearbooks.ws.BookService;
import com.nearsoft.nearbooks.ws.bodies.RequestBody;
import com.nearsoft.nearbooks.ws.responses.AvailabilityResponse;
import com.nearsoft.nearbooks.ws.responses.MessageResponse;

import java.io.IOException;
import java.util.List;

import retrofit2.Response;
import rx.Observable;
import rx.observers.TestSubscriber;

/**
 * <a href="http://d.android.com/tools/testing/testing_android.html">Testing Fundamentals</a>
 */
public class ApplicationTest extends ApplicationTestCase<NearbooksApplication> {

    private BookService mBookService;

    public ApplicationTest() {
        super(NearbooksApplication.class);
    }

    @Override
    protected void setUp() throws Exception {
        super.setUp();
        getApplication();
        SystemClock.sleep(100);
        NearbooksApplicationComponent nearbooksApplicationComponent = NearbooksApplication
                .getNearbooksApplicationComponent();
        mBookService = nearbooksApplicationComponent.providesBookService();
    }

    public void testBookService() throws IOException {
        Observable<List<Book>> observable = mBookService.getAllBooks();

        TestSubscriber<List<Book>> testSubscriber = new TestSubscriber<>();
        observable.subscribe(testSubscriber);

        testSubscriber.assertNoErrors();
        List<List<Book>> responses = testSubscriber.getOnNextEvents();
        assertNotNull(responses);
        assertEquals(1, responses.size());

        List<Book> books = responses.get(0);
        assertFalse(books.isEmpty());
    }

    public void testBookModel() {
        Book book = new Book();
        book.setId("123");
        book.setAuthor("epool");
        book.setAvailable(true);
        book.setTitle("Title");
        book.setReleaseYear(2015);
        book.setNumberOfCopies(1);
        book.setNumberOfDaysAllowedForBorrowing(7);
        if (!book.exists()) {
            book.save();
        }

        Book book1 = BookModel.findByBookId("123");

        assertTrue(book.exists());
        assertEquals(book, book1);
    }

    public void testBookAvailability() throws IOException {
        Observable<Response<AvailabilityResponse>> observable = mBookService
                .getBookAvailability("0321534468-0");

        TestSubscriber<Response<AvailabilityResponse>> testSubscriber = new TestSubscriber<>();
        observable.subscribe(testSubscriber);

        testSubscriber.assertNoErrors();
        List<Response<AvailabilityResponse>> responses = testSubscriber.getOnNextEvents();
        assertNotNull(responses);
        assertEquals(1, responses.size());

        Response<AvailabilityResponse> response = responses.get(0);
        assertNotNull(response);
        assertTrue(response.isSuccess());
        assertNotNull(response.body());
    }

    public void testRequestBookToBorrow() throws IOException {
        RequestBody requestBody = new RequestBody();
        requestBody.setQrCode("0321534468-0");
        requestBody.setUserEmail("epool@nearsoft.com");
        Observable<Response<Borrow>> observable = mBookService.requestBookToBorrow(requestBody);

        TestSubscriber<Response<Borrow>> testSubscriber = new TestSubscriber<>();
        observable.subscribe(testSubscriber);

        testSubscriber.assertNoErrors();
        List<Response<Borrow>> responses = testSubscriber.getOnNextEvents();
        assertNotNull(responses);
        assertEquals(1, responses.size());

        Response<Borrow> response = responses.get(0);
        assertNotNull(response);

        if (response.isSuccess()) {
            Borrow borrow = response.body();
            assertNotNull(borrow);
        } else {
            String errorMessage = ErrorUtil.parseError(response);
            assertNotNull(errorMessage);
        }
    }

    public void testCheckInBook() throws IOException {
        RequestBody requestBody = new RequestBody();
        requestBody.setQrCode("0321534468-1");
        requestBody.setUserEmail("epool@nearsoft.com");
        Observable<Response<MessageResponse>> observable = mBookService.checkInBook(requestBody);

        TestSubscriber<Response<MessageResponse>> testSubscriber = new TestSubscriber<>();
        observable.subscribe(testSubscriber);

        testSubscriber.assertNoErrors();
        List<Response<MessageResponse>> responses = testSubscriber.getOnNextEvents();
        assertNotNull(responses);
        assertEquals(1, responses.size());

        Response<MessageResponse> response = responses.get(0);
        assertNotNull(response);

        if (response.isSuccess()) {
            MessageResponse messageResponse = response.body();
            assertNotNull(messageResponse);
        } else {
            MessageResponse messageResponse = ErrorUtil.parseError(MessageResponse.class, response);
            assertNotNull(messageResponse);
        }
    }

    public void testCheckOutBook() throws IOException {
        RequestBody requestBody = new RequestBody();
        requestBody.setQrCode("0321534468-1");
        requestBody.setUserEmail("epool@nearsoft.com");
        Observable<Response<MessageResponse>> observable = mBookService.checkOutBook(requestBody);

        TestSubscriber<Response<MessageResponse>> testSubscriber = new TestSubscriber<>();
        observable.subscribe(testSubscriber);

        testSubscriber.assertNoErrors();
        List<Response<MessageResponse>> responses = testSubscriber.getOnNextEvents();
        assertNotNull(responses);
        assertEquals(1, responses.size());

        Response<MessageResponse> response = responses.get(0);
        assertNotNull(response);

        if (response.isSuccess()) {
            MessageResponse messageResponse = response.body();
            assertNotNull(messageResponse);
        } else {
            MessageResponse messageResponse = ErrorUtil.parseError(MessageResponse.class, response);
            assertNotNull(messageResponse);
        }
    }

}

File: app/src/androidTest/java/com/nearsoft/nearbooks/ApplicationTest.kt
package com.nearsoft.nearbooks

import android.os.SystemClock
import android.test.ApplicationTestCase
import com.nearsoft.nearbooks.models.realm.Book
import com.nearsoft.nearbooks.models.realm.Borrow
import com.nearsoft.nearbooks.util.ErrorUtil
import com.nearsoft.nearbooks.ws.BookService
import com.nearsoft.nearbooks.ws.GoogleBooksService
import com.nearsoft.nearbooks.ws.bodies.RequestBody
import com.nearsoft.nearbooks.ws.responses.AvailabilityResponse
import com.nearsoft.nearbooks.ws.responses.MessageResponse
import com.nearsoft.nearbooks.ws.responses.googlebooks.GoogleBooksSearchResponse
import io.realm.Realm
import retrofit2.Response
import rx.observers.TestSubscriber
import java.io.IOException

/**
 * [Testing Fundamentals](http://d.android.com/tools/testing/testing_android.html)
 */
class ApplicationTest : ApplicationTestCase<NearbooksApplication>(NearbooksApplication::class.java) {

    private var mBookService: BookService? = null
    private var mGoogleBooksService: GoogleBooksService? = null

    @Throws(Exception::class)
    override fun setUp() {
        super.setUp()
        application
        SystemClock.sleep(100)
        val nearbooksApplicationComponent = NearbooksApplication.applicationComponent
        mBookService = nearbooksApplicationComponent.provideBookService()
        mGoogleBooksService = nearbooksApplicationComponent.provideGoogleBooksService()
    }

    @Throws(IOException::class)
    fun testBookService() {
        val observable = mBookService!!.getAllBooksRx()

        val testSubscriber = TestSubscriber<List<Book>>()
        observable.subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        val responses = testSubscriber.onNextEvents
        assertNotNull(responses)
        assertEquals(1, responses.size)

        val books = responses[0]
        assertFalse(books.isEmpty())
    }

    fun testBookModel() {
        val book = com.nearsoft.nearbooks.models.realm.Book()
        book.id = "123"
        book.author = "epool"
        book.isAvailable = true
        book.title = "Title"
        book.releaseYear = "2015"
        book.numberOfCopies = 1
        book.numberOfDaysAllowedForBorrowing = 7

        val realm = Realm.getDefaultInstance()

        realm.executeTransaction { realm.copyToRealmOrUpdate(book) }

        val book1 = realm.where(com.nearsoft.nearbooks.models.realm.Book::class.java)
                .equalTo(Book.ID, "123")
                .findFirst()

        assertNotNull(book1)
        assertEquals(book.id, book1.id)

        realm.executeTransaction { book1.deleteFromRealm() }

        realm.close()
    }

    @Throws(IOException::class)
    fun testBookAvailability() {
        val observable = mBookService!!.getBookAvailability("0321534468-0")

        val testSubscriber = TestSubscriber<Response<AvailabilityResponse>>()
        observable.subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        val responses = testSubscriber.onNextEvents
        assertNotNull(responses)
        assertEquals(1, responses.size)

        val response = responses[0]
        assertNotNull(response)
        assertTrue(response.isSuccessful)
        assertNotNull(response.body())
    }

    @Throws(IOException::class)
    fun testRequestBookToBorrow() {
        val requestBody = RequestBody()
        requestBody.qrCode = "0321534468-0"
        requestBody.userEmail = "epool@nearsoft.com"
        val observable = mBookService!!.requestBookToBorrow(requestBody)

        val testSubscriber = TestSubscriber<Response<Borrow>>()
        observable.subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        val responses = testSubscriber.onNextEvents
        assertNotNull(responses)
        assertEquals(1, responses.size)

        val response = responses[0]
        assertNotNull(response)

        if (response.isSuccessful) {
            val borrow = response.body()
            assertNotNull(borrow)
        } else {
            val errorMessage = ErrorUtil.parseError(response)
            assertNotNull(errorMessage)
        }
    }

    @Throws(IOException::class)
    fun testCheckInBook() {
        val requestBody = RequestBody()
        requestBody.qrCode = "0321534468-1"
        requestBody.userEmail = "epool@nearsoft.com"
        val observable = mBookService!!.checkInBook(requestBody)

        val testSubscriber = TestSubscriber<Response<MessageResponse>>()
        observable.subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        val responses = testSubscriber.onNextEvents
        assertNotNull(responses)
        assertEquals(1, responses.size)

        val response = responses[0]
        assertNotNull(response)

        if (response.isSuccessful) {
            val messageResponse = response.body()
            assertNotNull(messageResponse)
        } else {
            val messageResponse = ErrorUtil.parseError(MessageResponse::class.java, response)
            assertNotNull(messageResponse)
        }
    }

    @Throws(IOException::class)
    fun testCheckOutBook() {
        val requestBody = RequestBody()
        requestBody.qrCode = "0321534468-1"
        requestBody.userEmail = "epool@nearsoft.com"
        val observable = mBookService!!.checkOutBook(requestBody)

        val testSubscriber = TestSubscriber<Response<MessageResponse>>()
        observable.subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        val responses = testSubscriber.onNextEvents
        assertNotNull(responses)
        assertEquals(1, responses.size)

        val response = responses[0]
        assertNotNull(response)

        if (response.isSuccessful) {
            val messageResponse = response.body()
            assertNotNull(messageResponse)
        } else {
            val messageResponse = ErrorUtil.parseError(MessageResponse::class.java, response)
            assertNotNull(messageResponse)
        }
    }

    @Throws(IOException::class)
    fun testGoogleBooksFindByIsbn() {
        val observable = mGoogleBooksService!!.findBooksByIsbn("9780321534460")

        val testSubscriber = TestSubscriber<Response<GoogleBooksSearchResponse>>()
        observable.subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        val responses = testSubscriber.onNextEvents
        assertNotNull(responses)
        assertEquals(1, responses.size)

        val response = responses[0]
        assertNotNull(response)
        assertTrue(response.isSuccessful)
        assertNotNull(response.body())
        assertTrue(response.body().totalItems > 0)
    }

}